Fixed #7164 -- Irrelevant layers are not saved.

jorgecarleitao opened the pull request:
This fixes #7164 by ignoring (in `get_config`) input and output layers that are not used by the model. This is consistent with the behavior we use for intermediary layers.

I also tried improving the readability of this part of the code (second commit) that makes no semantic change.

fchollet commented:
That's reasonable, but your code should respect the codebase style, in particular:
- docstring format
- reasonably sized lines

joeyearsley commented:
Made the additions here: https://github.com/joeyearsley/keras/tree/issue7164 

Should I add a new PR?

jorgecarleitao commented:
@fchollet, I modified the last commit to include those changes and rebased the whole PR against master. Should be ready to merge.

Thanks @joeyearsley for pointing out what is meant with "reasonably sized lines". :)

joeyearsley commented:
@jorgecarleitao No problem, usually I have my IDE use a line length of 80 before wrapping, I think this is in PEP8 somewhere, I always forget were exactly.

fchollet commented on keras/engine/topology.py:
```diff
@@ -2312,18 +2328,21 @@ def get_config(self):
-                node_key = layer.name + '_ib-' + str(original_node_index)
-                if node_key in self.container_nodes:
+                if self._node_key(layer, original_node_index) in \
+                        self.container_nodes:
```
Don't use `\` for breaking lines, prefer using parentheses

fchollet commented on keras/engine/topology.py:
```diff
@@ -312,6 +312,19 @@ def __init__(self, **kwargs):
+    def _node_key(layer, node_index):
+        """Converts a layer and its index to a unique (immutable type) name.
+        This function is used internally with `self.container_nodes`.
+        # Arguments
```
Insert line break above

fchollet commented on keras/engine/topology.py:
```diff
@@ -312,6 +312,19 @@ def __init__(self, **kwargs):
+            layer: The layer.
+            node_index: The layer's position (e.g. via enumerate) in a list of
+                nodes.
+        # Returns
```
Insert line break above

jorgecarleitao commented:
All comments fixed, squashed commits and rebased against master. It should be ready to go.

fchollet commented on keras/engine/topology.py:
```diff
@@ -2290,18 +2308,21 @@ def get_config(self):
-                node_key = layer.name + '_ib-' + str(original_node_index)
-                if node_key in self.container_nodes:
+                if (self._node_key(layer, original_node_index) in
+                        self.container_nodes):
```
Style nit: avoid strange line breaking

```python
node_key = self._node_key(layer, original_node_index)
if node_key in self.container_nodes:
```

Also applicable in several other places in this PR. Please fix.

jorgecarleitao replied:
Thanks! Done.
